### PR TITLE
Wizard recipe: libgmime-v3.2.15

### PR DESCRIPTION
**Pull request by AlexKlo**
This pull request contains a new build recipe I built using the BinaryBuilder.jl wizard:

* Package name: libgmime
* Version: v3.2.15

@staticfloat please review and merge.

### CODE REVIEW

File: L/libgmime/build_tarballs.jl [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+# Collection of sources required to complete build
+sources = [
+    ArchiveSource("https://github.com/jstedfast/gmime/releases/download/$(version)/gmime-$(version).tar.xz", "84cd2a481a27970ec39b5c95f72db026722904a2ccf3fdbd57b280cf2d02b5c4")
+]
+
+# Bash recipe for building across all platforms
+script = raw"""
+cd $WORKSPACE/srcdir
+cd gmime-3.2.15/
```
Comment (giordano) [inline L/libgmime/build_tarballs.jl:16]:
```suggestion
cd $WORKSPACE/srcdir/gmime*
```

File: L/libgmime/build_tarballs.jl [hunk 2]
```diff
@@ -0,0 +1,48 @@
+# Note that this script can accept some limited command-line arguments, run
+# `julia build_tarballs.jl --help` to see a usage message.
+using BinaryBuilder, Pkg
+
+name = "libgmime"
```
Comment (giordano) [inline L/libgmime/build_tarballs.jl:5]: https://docs.binarybuilder.org/stable/building/#Name
```suggestion
name = "gmime"
```

File: L/libgmime/build_tarballs.jl [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+cd $WORKSPACE/srcdir
+cd gmime-3.2.15/
+echo '#define ICONV_ISO_INT_FORMAT "iso-%u-%u"' > iconv-detect.h
+echo '#define ICONV_ISO_STR_FORMAT "iso-%u-%s"' >> iconv-detect.h
+echo '#define ICONV_SHIFT_JIS "shift-jis"' >> iconv-detect.h
+echo '#define ICONV_10646 "UCS-4BE"' >> iconv-detect.h
+
+export LDFLAGS="$LDFLAGS -liconv"
+CMAKE_EXE_LINKER_FLAGS="-Wl,--copy-dt-needed-entries"
+export CMAKE_EXE_LINKER_FLAGS
```
Comment (giordano) [inline L/libgmime/build_tarballs.jl:24]: Can you please elaborate why this is needed?
```suggestion
export CMAKE_EXE_LINKER_FLAGS="-Wl,--copy-dt-needed-entries"
```

File: L/libgmime/build_tarballs.jl [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+# Bash recipe for building across all platforms
+script = raw"""
+cd $WORKSPACE/srcdir
+cd gmime-3.2.15/
+echo '#define ICONV_ISO_INT_FORMAT "iso-%u-%u"' > iconv-detect.h
+echo '#define ICONV_ISO_STR_FORMAT "iso-%u-%s"' >> iconv-detect.h
+echo '#define ICONV_SHIFT_JIS "shift-jis"' >> iconv-detect.h
+echo '#define ICONV_10646 "UCS-4BE"' >> iconv-detect.h
+
+export LDFLAGS="$LDFLAGS -liconv"
```
Comment (giordano) [inline L/libgmime/build_tarballs.jl:22]: Why is this needed?
```suggestion
export LDFLAGS="-liconv"
```

File: L/libgmime/build_tarballs.jl [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+
+# The products that we will ensure are always built
+products = [
+    LibraryProduct(["libgmime-3", "libgmime-3.0"], :libgmime)
+]
+
+# Dependencies that must be installed before this package can be built
+dependencies = [
+    Dependency(PackageSpec(name="Glib_jll", uuid="7746bdde-850d-59dc-9ae8-88ece973131d"))
```
Comment (giordano) [inline L/libgmime/build_tarballs.jl:43]: #3024
```suggestion
    Dependency(PackageSpec(name="Glib_jll", uuid="7746bdde-850d-59dc-9ae8-88ece973131d"); compat="2.68.1")
```

File: L/libgmime/build_tarballs.jl [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+    Dependency(PackageSpec(name="Libiconv_jll", uuid="94ce4f54-9a6c-5748-9c1c-f9c7231a4531"))
```
Comment (giordano) [inline L/libgmime/build_tarballs.jl:44]: Is this really needed?  In particular, on all platforms?

File: L/libgmime/build_tarballs.jl (hunk 2 above)
Comment (giordano) [reply L/libgmime/build_tarballs.jl]: After you rename this you also need to move the file to a directory with a matching name.

File: L/libgmime/build_tarballs.jl [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+# The products that we will ensure are always built
+products = [
+    LibraryProduct(["libgmime-3", "libgmime-3.0"], :libgmime)
+]
+
+# Dependencies that must be installed before this package can be built
+dependencies = [
+    Dependency(PackageSpec(name="Glib_jll", uuid="7746bdde-850d-59dc-9ae8-88ece973131d"); compat="2.68.1")
+    BuildDependency(PackageSpec(name="Libiconv_jll", uuid="94ce4f54-9a6c-5748-9c1c-f9c7231a4531"))
```
Comment (giordano) [inline L/libgmime/build_tarballs.jl:45]: I think build were passing without libiconv, no?

Comment (AlexKlo) [reply L/libgmime/build_tarballs.jl:45]: No, I tried without libiconv, it was the same error.

Comment (giordano) [reply L/libgmime/build_tarballs.jl:45]: I don't think so: https://buildkite.com/julialang/yggdrasil/builds/13256#0191f68e-4084-4092-877a-6c55f172df2b

Comment (giordano) [reply L/libgmime/build_tarballs.jl:45]: Ah, but libiconv is pulled in by glib anyway.

File: L/libgmime/build_tarballs.jl (hunk 3 above)
Comment (giordano) [reply L/libgmime/build_tarballs.jl]: Please answer the question.

File: L/libgmime/build_tarballs.jl (hunk 4 above)
Comment (giordano) [reply L/libgmime/build_tarballs.jl]: Please answer the question

File: L/libgmime/build_tarballs.jl (hunk 2 above)
Comment (AlexKlo) [reply L/libgmime/build_tarballs.jl]: Sorry, I don't understand what exactly you mean

Comment (giordano) [reply L/libgmime/build_tarballs.jl]: The file is in `L/libgmime/build_tarballs.jl`, it has to be moved to `G/gmime/build_tarballs.jl`, otherwise jobs will keep failing forever.

File: L/libgmime/build_tarballs.jl (hunk 3 above)
Comment (AlexKlo) [reply L/libgmime/build_tarballs.jl]: This fixes issue with [error adding symbols: DSO missing from command line](https://stackoverflow.com/questions/19901934/libpthread-so-0-error-adding-symbols-dso-missing-from-command-line)

File: L/libgmime/build_tarballs.jl (hunk 4 above)
Comment (AlexKlo) [reply L/libgmime/build_tarballs.jl]: same as above

File: L/libgmime/build_tarballs.jl (hunk 7 above)
Comment (giordano) [reply L/libgmime/build_tarballs.jl]: Two are the options:

1. libiconv is needed and then you must added it as explicit dependency
2. libiconv is not needed and then it shouldn't be an explicit dependency

Having a situation with "libiconv is needed" and "not listed as dependency" is not good.

File: L/libgmime/build_tarballs.jl (hunk 3 above)
Comment (giordano) [reply L/libgmime/build_tarballs.jl]: I still don't get this one.  Can you please provide the full compiler invocation and related error message?

Comment (AlexKlo) [reply L/libgmime/build_tarballs.jl]: Hello,
Here is error message:
```
/opt/x86_64-linux-gnu/bin/../lib/gcc/x86_64-linux-gnu/8.1.0/../../../../x86_64-linux-gnu/bin/ld: charset-map.o: undefined reference to symbol 'libiconv'
/opt/x86_64-linux-gnu/bin/../lib/gcc/x86_64-linux-gnu/8.1.0/../../../../x86_64-linux-gnu/bin/ld: /workspace/destdir/lib/libiconv.so.2: error adding symbols: DSO missing from command line
collect2: error: ld returned 1 exit status
make[3]: *** [Makefile:817: charset-map] Error 1
make[3]: *** Waiting for unfinished jobs....
```

Documentstion also [says](https://fossies.org/linux/gmime/docs/reference/html/gmime-building.html#:~:text=The%20GNU%20libiconv%20library%20is%20needed%20to%20build%20GLib%20and%20GMime%20if%20your%20system%20doesn%27t%20already%20have%20the%20iconv()%20function%20for%20doing%20conversion%20between%20character%20encodings.%20Most%20modern%20systems%20should%20have%20iconv().) that `libiconv` needed if the system doesn't have `iconv`

File: L/libgmime/build_tarballs.jl (hunk 4 above)
Comment (AlexKlo) [reply L/libgmime/build_tarballs.jl]: Without adding iconv-detect.h there is an error:
```
gmime-charset.c:49:10: fatal error: iconv-detect.h: No such file or directory
 #include "iconv-detect.h"
          ^~~~~~~~~~~~~~~~
compilation terminated.
make[3]: *** [Makefile:917: gmime-charset.lo] Error 1
make[3]: *** Waiting for unfinished jobs....
```

Comment (AlexKlo) [reply L/libgmime/build_tarballs.jl]: Without `ac_cv_have_iconv_detect_h=yes` key
```
checking preferred charset formats for system iconv... cannot run test program while cross compiling
configure: error: Copy target system's iconv-detect.h to . and configure with ac_cv_have_iconv_detect_h=yes
```

File: L/libgmime/build_tarballs.jl (hunk 7 above)
Comment (AlexKlo) [reply L/libgmime/build_tarballs.jl]: libiconv is needed, so I think it's a complete script now

Comment (AlexKlo) [reply L/libgmime/build_tarballs.jl]: @giordano сould you please continue the review?

File: G/gmime/build_tarballs.jl [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+
+# The products that we will ensure are always built
+products = [
+    LibraryProduct(["libgmime-3", "libgmime-3.0"], :libgmime)
+]
+
+# Dependencies that must be installed before this package can be built
+dependencies = [
+    Dependency(PackageSpec(name="Glib_jll", uuid="7746bdde-850d-59dc-9ae8-88ece973131d"); compat="2.68.1")
+    BuildDependency(PackageSpec(name="Libiconv_jll", uuid="94ce4f54-9a6c-5748-9c1c-f9c7231a4531"))
```
Comment (giordano) [inline G/gmime/build_tarballs.jl:45]: https://github.com/JuliaPackaging/Yggdrasil/pull/9427#discussion_r1760342714

File: G/gmime/build_tarballs.jl [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+# Bash recipe for building across all platforms
+script = raw"""
+cd $WORKSPACE/srcdir/gmime*
+echo '#define ICONV_ISO_INT_FORMAT "iso-%u-%u"' > iconv-detect.h
+echo '#define ICONV_ISO_STR_FORMAT "iso-%u-%s"' >> iconv-detect.h
+echo '#define ICONV_SHIFT_JIS "shift-jis"' >> iconv-detect.h
+echo '#define ICONV_10646 "UCS-4BE"' >> iconv-detect.h
+
+export LDFLAGS="-liconv"
+export CMAKE_EXE_LINKER_FLAGS="-Wl,--copy-dt-needed-entries"
```
Comment (giordano) [inline G/gmime/build_tarballs.jl:22]: I'm still not clear why this is necessary, we never had any package needing something like this.

File: G/gmime/build_tarballs.jl [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+
+# Bash recipe for building across all platforms
+script = raw"""
+cd $WORKSPACE/srcdir/gmime*
+echo '#define ICONV_ISO_INT_FORMAT "iso-%u-%u"' > iconv-detect.h
+echo '#define ICONV_ISO_STR_FORMAT "iso-%u-%s"' >> iconv-detect.h
+echo '#define ICONV_SHIFT_JIS "shift-jis"' >> iconv-detect.h
+echo '#define ICONV_10646 "UCS-4BE"' >> iconv-detect.h
+
+export LDFLAGS="-liconv"
```
Comment (giordano) [inline G/gmime/build_tarballs.jl:21]: The fact that you need to do this manually frankly looks to me an upstream bug.

File: G/gmime/build_tarballs.jl (hunk 8 above)
Comment (AlexKlo) [reply G/gmime/build_tarballs.jl:45]: libiconv is needed, so what BuildDependency added

File: G/gmime/build_tarballs.jl (hunk 9 above)
Comment (AlexKlo) [reply G/gmime/build_tarballs.jl:22]: There is building error without it
image

File: G/gmime/build_tarballs.jl (hunk 10 above)
Comment (AlexKlo) [reply G/gmime/build_tarballs.jl:21]: This is a message after `./configure --prefix=${prefix}   --build=${MACHTYPE}   --host=${target}`
```
checking preferred charset formats for system iconv... cannot run test program while cross compiling
configure: error: Copy target system's iconv-detect.h to . and configure with ac_cv_have_iconv_detect_h=yes
```
So what added iconv-detect.h manually

File: G/gmime/build_tarballs.jl (hunk 8 above)
Comment (giordano) [reply G/gmime/build_tarballs.jl:45]: It ***must*** be a ***runtime*** dependency (i.e. `Dependency`, not `BuildDependency`) if you're linking to it.

File: G/gmime/build_tarballs.jl (hunk 9 above)
Comment (giordano) [reply G/gmime/build_tarballs.jl:22]: I can compile for `x86_64-linux-gnu` without this line.
```suggestion
```

File: G/gmime/build_tarballs.jl (hunk 10 above)
Comment (giordano) [reply G/gmime/build_tarballs.jl:21]: And in fact I can compile without and explicit `LDFLAGS="-liconv"`

**Comment by giordano** [comment]
```diff
diff --git a/G/gmime/build_tarballs.jl b/G/gmime/build_tarballs.jl
index 59a4e099b..e453a2f04 100644
--- a/G/gmime/build_tarballs.jl
+++ b/G/gmime/build_tarballs.jl
@@ -7,7 +7,8 @@ version = v"3.2.15"
 
 # Collection of sources required to complete build
 sources = [
-    ArchiveSource("https://github.com/jstedfast/gmime/releases/download/$(version)/gmime-$(version).tar.xz", "84cd2a481a27970ec39b5c95f72db026722904a2ccf3fdbd57b280cf2d02b5c4")
+    ArchiveSource("https://github.com/jstedfast/gmime/releases/download/$(version)/gmime-$(version).tar.xz",
+                  "84cd2a481a27970ec39b5c95f72db026722904a2ccf3fdbd57b280cf2d02b5c4")
 ]
 
 # Bash recipe for building across all platforms
@@ -18,12 +19,16 @@ echo '#define ICONV_ISO_STR_FORMAT "iso-%u-%s"' >> iconv-detect.h
 echo '#define ICONV_SHIFT_JIS "shift-jis"' >> iconv-detect.h
 echo '#define ICONV_10646 "UCS-4BE"' >> iconv-detect.h
 
-export LDFLAGS="-liconv"
-export CMAKE_EXE_LINKER_FLAGS="-Wl,--copy-dt-needed-entries"
+if [[ "${target}" == *-apple-* ]]; then
+   # Help the linker prefer our libiconv over system's libiconv.
+   export LDFLAGS="-L${libdir}"
+fi
 
 ./configure --prefix=${prefix} \
             --build=${MACHTYPE} \
             --host=${target} \
+            --enable-static=no \
+            --with-libiconv=gnu \
             ac_cv_have_iconv_detect_h=yes
 make -j${nproc}
 make install
@@ -31,7 +36,7 @@ make install
 
 # These are the platforms we will build for by default, unless further
 # platforms are passed in on the command line
-platforms = supported_platforms(exclude=[Platform("aarch64", "FreeBSD")])
+platforms = supported_platforms(; exclude=p -> Sys.isfreebsd(p) && arch(p) == "aarch64")
 
 
 # The products that we will ensure are always built
@@ -42,7 +47,7 @@ products = [
 # Dependencies that must be installed before this package can be built
 dependencies = [
     Dependency(PackageSpec(name="Glib_jll", uuid="7746bdde-850d-59dc-9ae8-88ece973131d"); compat="2.68.1")
-    BuildDependency(PackageSpec(name="Libiconv_jll", uuid="94ce4f54-9a6c-5748-9c1c-f9c7231a4531"))
+    Dependency(PackageSpec(name="Libiconv_jll", uuid="94ce4f54-9a6c-5748-9c1c-f9c7231a4531"))
 ]
 
 # Build the tarballs, and possibly a `build.jl` as well.
```
This works for me on all the platforms I tried.  No need to use obscure linker flags, nor explicit `-liconv`.

File: G/gmime/build_tarballs.jl (hunk 9 above)
Comment (giordano) [reply G/gmime/build_tarballs.jl:22]: As demonstrated at https://github.com/JuliaPackaging/Yggdrasil/pull/9427#issuecomment-2361981292 this flags is not needed at all.

**Comment by AlexKlo** [comment]
Not script looks better! Thank you very much @giordano , you really helped me!

**Comment by giordano** [comment]
Thanks for your contribution!
